tools/aux/tpm2_print.c

```diff
@@ -0,0 +1,314 @@
+    if (magic != TPM2_GENERATED_VALUE) {
+        LOG_ERR("%s: Bad magic", file_path);
+        goto error;
+    }
+
+    UINT16 type;
+    res = files_read_bytes(fd, (UINT8*)&type, sizeof(UINT16));
+    if (!res) {
+        goto read_error;
+    }
+    tpm2_tool_output("type: ");
+    tpm2_util_hexdump((const UINT8*)&type, sizeof(UINT16), true);
+    tpm2_tool_output("\n");
+    type = tpm2_util_ntoh_16(type); // finally, convert endianness
+
+    tpm2_tool_output("qualifiedSigner: ");
+    res = tpm2_util_print_tpm2b_file(fd);
+    tpm2_tool_output("\n");
+    if (!res) {
+        goto read_error;
+    }
+
+    tpm2_tool_output("extraData: ");
+    res = tpm2_util_print_tpm2b_file(fd);
+    tpm2_tool_output("\n");
+    if (!res) {
+        goto read_error;
+    }
+
+    tpm2_tool_output("clockInfo:\n");
+    res = print_clock_info(fd, 1);
+    if (!res) {
+        goto error;
+    }
+
+    tpm2_tool_output("firmwareVersion: ");
+    res = tpm2_util_hexdump_file(fd, sizeof(UINT64), true);
+    tpm2_tool_output("\n");
+    if (!res) {
+        goto read_error;
+    }
+
+    switch(type) {
+    case TPM2_ST_ATTEST_QUOTE:
+        tpm2_tool_output("attested:\n");
+        print_yaml_indent(1);
+        tpm2_tool_output("quote:\n");
+        res = print_TPMS_QUOTE_INFO(fd, 2);
+        if (!res) {
+            goto error;
+        }
+        break;
+
+    default:
+        LOG_ERR("%s: Cannot print unsupported type 0x%x", file_path, (unsigned int)type);
+        goto error;
+    }
+
+    return true;
+
+read_error:
+    LOG_ERR("%s: File too short", file_path);
+error:
+    return false;
+}
+
+static bool on_option(char key, char *value) {
+    switch (key) {
+    case 't':
+        if (strcmp(value, "TPMS_ATTEST") != 0) {
+            LOG_ERR("Invalid type specified. Only TPMS_ATTEST is presently supported.\n");
+            return false;
+        }
+        file_type = file_type_TPMS_ATTEST;
+        break;
+    case 'f':
+        file_path = value;
+        break;
+    default:
+        LOG_ERR("Invalid option %c\n", key);
+        return false;
+    }
+
+    return true;
+}
+
+bool tpm2_tool_onstart(tpm2_options **opts) {
+    static const struct option topts[] = {
+        { "type",        required_argument, NULL, 't' },
+        { "file",        optional_argument, NULL, 'f' },
+    };
+
+    *opts = tpm2_options_new("t:f:", ARRAY_LEN(topts), topts,
+        on_option, NULL);
+
+    return *opts != NULL;
+}
+
+int tpm2_tool_onrun(TSS2_SYS_CONTEXT *sapi_context, tpm2_option_flags flags) {
+    UNUSED(sapi_context);
+    UNUSED(flags);
+
+    bool res;
+    FILE* fd = stdin;
+
+    if (file_path && *file_path) {
+        LOG_INFO("Reading from file_path=%s", file_path);
+        fd = fopen(file_path, "rb");
+        if(!fd) {
+            LOG_ERR("%s: Could not open file", file_path);
+            return 1;
+        }
+    }
+    else {
+        LOG_INFO("Reading from stdin");
+    }
+
+    switch (file_type) {
+    case file_type_TPMS_ATTEST:
```

Comment:
You could check this before even attempting to read the file.... this way you can fail early before you even have to worry about closing the file.